extensions/amp-app-banner/0.1/amp-app-banner.js

```diff
@@ -0,0 +1,405 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {user, dev, rethrowAsync} from '../../../src/log';
+import {platform} from '../../../src/platform';
+import {viewerFor} from '../../../src/viewer';
+import {viewportFor} from '../../../src/viewport';
+import {vsyncFor} from '../../../src/vsync';
+import {CSS} from '../../../build/amp-app-banner-0.1.css';
+import {documentInfoFor} from '../../../src/document-info';
+import {xhrFor} from '../../../src/xhr';
+import {assertHttpsUrl} from '../../../src/url';
+import {isExperimentOn} from '../../../src/experiments';
+import {removeElement, openWindowDialog} from '../../../src/dom';
+import {storageFor} from '../../../src/storage';
+import {timer} from '../../../src/timer';
+import {parseUrl} from '../../../src/url';
+
+
+const TAG = 'amp-app-banner';
+
+
+/**
+ * @private visible for testing.
+ */
+export class AbstractAppBanner extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.CONTAINER;
+  }
+
+  /** @protected */
+  setupOpenLink_(openInAppUrl, installAppUrl) {
+    const openLink = this.element.querySelector('a[open-link]');
+    openLink.addEventListener('click', () => {
+      this.openLinkClicked_(openInAppUrl, installAppUrl);
+    });
+  }
+
+
+  /** @private */
+  openLinkClicked_(openInAppUrl, installAppUrl) {
+    timer.delay(() => {
+      this.redirectTopLocation_(installAppUrl);
+    }, 1500);
+    openWindowDialog(this.win, openInAppUrl, '_top');
+  }
+
+
+  /** @private */
+  redirectTopLocation_(link) {
+    this.win.top.location.replace(link);
+  }
+
+  /**
+   * Creates and append a close button.
+   * @protected
+   */
+  addDismissButton_() {
+    const dismissButton = this.win.document.createElement('button');
+    dismissButton.classList.add('amp-app-banner-dismiss-button');
+    dismissButton.setAttribute('aria-label',
+        this.element.getAttribute('data-dismiss-button-aria-label') ||
+        'Dismiss');
+    const boundOnDismissButtonClick = this.onDismissButtonClick_.bind(this);
+    dismissButton.addEventListener('click', boundOnDismissButtonClick);
+    this.element.appendChild(dismissButton);
+  }
+
+  /**
+   * Dismisses the app banner and persist it.
+   * @protected
+   */
+  onDismissButtonClick_() {
+    vsyncFor(this.win).run({
```

Comment:
mutate doesn't support state passing yet. Justin opened an issue to add that in #3776